unicorn/include/cocaine/api/unicorn.hpp

```diff
@@ -0,0 +1,195 @@
+     **/
+    struct response {
+        typedef result_of<io::unicorn::put>::type put;
+        typedef result_of<io::unicorn::create>::type create;
+        typedef result_of<io::unicorn::del>::type del;
+        typedef result_of<io::unicorn::increment>::type increment;
+        typedef result_of<io::unicorn::get>::type get;
+        typedef result_of<io::unicorn::subscribe>::type subscribe;
+        typedef result_of<io::unicorn::children_subscribe>::type children_subscribe;
+        typedef result_of<io::unicorn::lock>::type lock;
+    };
+
+    /**
+     * Typedefs for used writable helpers
+     * These provide future-like functionality.
+     * Either value can be written inside via call to write,
+     * or an error can be signaled via call to abort.
+     * @see cocaine::unicorn::writable_adapter_base_t
+     **/
+    struct writable_ptr {
+        typedef unicorn::writable_helper<response::put>::ptr put;
+        typedef unicorn::writable_helper<response::create>::ptr create;
+        typedef unicorn::writable_helper<response::del>::ptr del;
+        typedef unicorn::writable_helper<response::increment>::ptr increment;
+        typedef unicorn::writable_helper<response::get>::ptr get;
+        typedef unicorn::writable_helper<response::subscribe>::ptr subscribe;
+        typedef unicorn::writable_helper<response::children_subscribe>::ptr children_subscribe;
+        typedef unicorn::writable_helper<response::lock>::ptr lock;
+    };
+
+    unicorn_t(context_t& context, const std::string& name, const dynamic_t& args);
+
+    /**
+     * Put value in unicorn.
+     * @param result writable to which will be result of operation written.
+     * @param path path to put value in with starting '/'.
+     * @param value value to put in path.
+     * @param version current version in unicorn. If it do not match - unicorn will return an error.
+     */
+    virtual
+    unicorn_scope_ptr
+    put(writable_ptr::put result,
+        const unicorn::path_t& path,
+        const unicorn::value_t& value,
+        unicorn::version_t version) = 0;
+
+    /**
+     * Get value from unicorn.
+     * @param result writable to which will be result of operation written.
+     * @param path path to read value from.
+     */
+    virtual
+    unicorn_scope_ptr
+    get(writable_ptr::get result,
+        const unicorn::path_t& path) = 0;
+
+    /**
+     * Create a node on specified path with specified value.
+     */
+    virtual
+    unicorn_scope_ptr
+    create(writable_ptr::create result,
+           const unicorn::path_t& path,
+           const unicorn::value_t& value,
+           bool ephemeral,
+           bool sequence) = 0;
+
+    unicorn_scope_ptr
+    create_default(writable_ptr::create result,
+                   const unicorn::path_t& path,
+                   const unicorn::value_t& value)
+    {
+        return create(std::move(result), std::move(path), std::move(value), false, false);
```

Comment:
You cast const lvalue reference to mutable rvalue. Are you sure that it's exactly what you want?